loki/ir/transformer.py

```diff
@@ -107,10 +98,12 @@ def _rebuild(self, o, children, **args):
         args_frozen = o.args_frozen
         args_frozen.update(args)
         if self.invalidate_source and 'source' in args_frozen:
-            if any(isinstance(child, Node) for child in flatten(children)):
-                child_has_no_source = [getattr(i, 'source', None) is None for i in flatten(children)]
-                if any(child_has_no_source) or len(child_has_no_source) != len(flatten(o.children)):
-                    return self._rebuild_without_source(o, children, **args_frozen)
+            # If any child node has been invalidated, mark this node as invalid too
+            if any(isinstance(c, Node) and not is_source_valid(c) for c in flatten(children)):
+                if is_source_valid(args_frozen.get('source')):
```

Comment:
Might want to reverse the nesting of the conditionals here to save the loop through children